retry/retry.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+// Package retry provides a framework for retrying actions.
+// It does not itself invoke the action to be retried, but
+// is intended to be used in a retry loop.
+//
+// The basic usage is as follows:
+//
+//	for a := someStrategy.Start(); a.Next(); {
+//		try()
+//	}
+//
+// See examples for details of suggested usage.
+package retry
+
+import (
+	"time"
+
+	"github.com/juju/utils/clock"
+)
+
+// Strategy is implemented by types that represent a retry strategy.
+type Strategy interface {
+	NewTimer(now time.Time) Timer
+}
+
+// Timer represents a source of timing events for a retry strategy.
+type Timer interface {
+	// NextSleep returns the length of time to sleep
+	// before the next retry. If no more attempts should
+	// be made it should return false, and the returned
+	// duration will be ignored.
+	//
+	// Note that NextSleep is called once after
+	// each iteration has completed, assuming the
+	// retry loop is continuing.
+	NextSleep(now time.Time) (time.Duration, bool)
+}
+
+// Attempt represents a running retry attempt.
+type Attempt struct {
+	clock   clock.Clock
+	stop    <-chan struct{}
+	timer   Timer
+	count   int
+	waited  bool
+	running bool
+}
+
+// Start begins a new sequence of attempts for the given strategy. If
+// clk is nil, clock.WallClock will be used. If a value is received on
+// stop while waiting, the attempt will be aborted.
+func Start(strategy Strategy, clk clock.Clock, stop <-chan struct{}) *Attempt {
+	if clk == nil {
+		clk = clock.WallClock
```

Comment:
I'd like this package to be useful for one-off programs too and requiring every such use
to have an explicit dependency on juju/utils/clock seems a bit wrong to me.
If people are using a mock clock, they'll pass it in here. If they're not, I don't think
typing "clock.WallClock" is going to change matters much.